python/thirdweb-ai/src/thirdweb_ai/services/storage.py

```diff
@@ -0,0 +1,176 @@
+
+        return files_data
+
+    @tool(
+        description="Upload a file, directory, or JSON data to IPFS. Stores any type on decentralized storage and returns an IPFS URI."
+    )
+    async def upload_to_ipfs(
+        self,
+        data: Annotated[
+            Any, "Data to upload: can be a file path, directory path, dict, dataclass, or BaseModel instance."
+        ],
+        chunk_size: Annotated[int, "Size of chunks to read from file."] = 8192,
+    ) -> str:
+        """
+        Upload data to IPFS and return the IPFS hash.
+
+        Supports:
+        - File paths (streams content)
+        - Directory paths (preserves directory structure)
+        - Dict objects (converted to JSON)
+        - Dataclass instances (converted to JSON)
+        - Pydantic BaseModel instances (converted to JSON)
+
+        Always uses streaming for file uploads to handle large files efficiently.
+        """
+        storage_url = f"{self.base_url}/upload"
+
+        # Handle JSON-serializable data types
+        if self._is_json_serializable(data):
+            json_content = self._convert_to_json(data)
+            files = {"file": ("data.json", BytesIO(json_content.encode()), "application/json")}
+            body = await self._async_post_file(storage_url, files)
+            return f"ipfs://{body['IpfsHash']}"
+
+        # Handle string paths to files or directories
+        if isinstance(data, str) and self._is_valid_path(data):
+            path = Path(data)
+
+            # Single file upload with streaming
+            if path.is_file():
+                content_type = mimetypes.guess_type(str(path))[0] or "application/octet-stream"
+
+                # Create a buffer to hold chunks for streaming upload
+                buffer = BytesIO()
+                async for chunk in async_read_file_chunks(path, chunk_size):
+                    buffer.write(chunk)
+
+                buffer.seek(0)  # Reset buffer position
+                files = {"file": (path.name, buffer, content_type)}
+                body = await self._async_post_file(storage_url, files)
+                return f"ipfs://{body['IpfsHash']}"
+
+            # Directory upload - preserve directory structure
+            if path.is_dir():
+                # Prepare all files from the directory with preserved structure
+                files_data = await self._prepare_directory_files(path, chunk_size)
+
+                if not files_data:
+                    raise ValueError(f"Directory is empty: {data}")
+
+                files_dict = {
+                    f"file{i}": (relative_path, buffer, content_type)
+                    for i, (relative_path, buffer, content_type) in enumerate(files_data)
+                }
+                body = await self._async_post_file(storage_url, files_dict)
+                return f"ipfs://{body['IpfsHash']}"
+
+            raise ValueError(f"Path exists but is neither a file nor a directory: {data}")
+
+        raise ValueError(
```

Comment:
i think in the `else` case, we can just encode it as bytes and upload it.

so that we can instruct the agent to "write the answer of 2+2 to ipfs", then it'd pass "4" to this function, and we'd just upload it as plain text.